lib/puppet/parser/functions/upcase.rb

```diff
@@ -17,22 +17,22 @@ module Puppet::Parser::Functions
   ) do |arguments|
 
     raise(Puppet::ParseError, "upcase(): Wrong number of arguments " +
-                                "given (#{arguments.size} for 1)") if arguments.size < 1
+                                "given (#{arguments.size} for 1)") if arguments.size != 1
 
     value = arguments[0]
 
-    unless value.is_a?(Array) || value.is_a?(String) || value.is_a?(Hash)
+    unless value.is_a?(Array) || value.is_a?(Hash) || value.respond_to?(:upcase)
       raise(Puppet::ParseError, 'upcase(): Requires an ' +
-                                  'array, string or hash to work with')
+                                  'array, hash or object that responds to upcase in order to work')
     end
 
     if value.is_a?(Array)
       # Numbers in Puppet are often string-encoded which is troublesome ...
```

Comment:
As long as it is passed as a String type it will work as it always has, that being said, with future parser it will fail if it is passed a Numeric type, but this was always going to be the case.